### DISCUSSION TITLE
Path forward for image-related functionality

### DISCUSSION THREAD

**Discussion by rootwork** [chipzoller/hugo-clarity#362] [Planning]
**Background:** Complexity around images and image-related functionality (like captioning) has grown significantly -- mostly from my own additions. This has led to a much larger surface area for bugs, and many changes/fixes generate new ones. Current bugs include #326, #337, https://github.com/chipzoller/hugo-clarity/issues/310#issuecomment-1133941933. There's also an issue with putting classes on images but not on wrapping `<figure>` elements (which makes it impossible to do things like constrain or float an image with its caption).

**Proposal:** Consider adapting one of the other comprehensive Hugo theme modules for images that are out there. Instead of being responsible for all their functionality as part of this theme, we'd leverage work other people are already doing, and instead spend our time on things specific to our theme, like the figure numbering.

**Examples:** Two Hugo-oriented projects I know of are [DFD image handling](https://github.com/danielfdickinson/image-handling-mod-hugo-dfd) and [hyas/images](https://github.com/h-enk/hyas-images). Both of these have even more features than what we've already come up with -- for instance, auto-generating `WebP` versions using Hugo Pipes, providing responsive styles (not serving larger images than a device needs), [LQIP/blur-up](https://github.com/aFarkas/lazysizes), and better performance/accessibility.

**Checklist:**
- [ ] Decide if this is a good path forward generally
- [ ] Decide on an image component, such as the two I mentioned or others that may be out there
- [ ] Determine how to incorporate the module in a way easiest for end users: Hugo module (DFD's method)? npm module (hyas's method)? git submodule? manually copying module files into a `vendor` dir?
- [ ] Determine how to integrate the functionality provided by the module into our own, e.g. how it intersects with figure numbering
- [ ] Set up tests for the major use cases: Markdown image without page bundles, Markdown image with page bundles, theme image (e.g. featured image) without page bundles, theme image with page bundles.
- [ ] Yank out all the image-related stuff now being handled by the module
- [ ] Map configuration options to our own and set default options (possibly setting some as opt-in) so as to minimize disruptions for current users of the theme
- [ ] Credit sources in docs etc.

What do you think @chipzoller @onweru? In particular wondering about any opinions on the third item above.

**Comment by rootwork** [comment]
Just checking in @chipzoller @onweru - There are pending bugs in the image handling (some reported, and some I've recently discovered myself) but I'm reluctant to invest a lot of time into fixing it if we might adopt something else.

I suppose another alternative to adopting one of the above modules is to roll back to a simpler form of image handling and allow users to add whatever they want on top of it.

One challenge is that we try to support images whether you're using Page Bundles or not -- most themes I've seen have picked supporting one or the other. A lot of the logic in our image partials is about trying to navigate that.

**Reply by chipzoller** [reply]
Do we have a concise listing of all such bugs and in one issue/location?
